src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs

```diff
@@ -149,45 +140,75 @@ internal static SafeSslHandle AllocateSslContext(SslProtocols protocols, SafeX50
                     SetSslCertificate(innerContext, certHandle!, certKeyHandle!);
                 }
 
-                if (sslAuthenticationOptions.IsServer && sslAuthenticationOptions.RemoteCertRequired)
-                {
-                    unsafe
-                    {
-                        Ssl.SslCtxSetVerify(innerContext, &VerifyClientCertificate);
-                    }
-                }
+            }
+            catch
+            {
+                innerContext.Dispose();
+                throw;
+            }
+
+            return innerContext;
+        }
+
+        // This essentially wraps SSL* SSL_new()
+        internal static SafeSslHandle AllocateSslHandle(SafeFreeSslCredentials credential, SslAuthenticationOptions sslAuthenticationOptions)
+        {
+            SafeSslHandle? context = null;
+            SafeSslContextHandle? sslCtx = null;
+            SafeSslContextHandle? innerContext = null;
+            SslProtocols protocols = sslAuthenticationOptions.EnabledSslProtocols;
+            bool cacheSslContext = sslAuthenticationOptions.EncryptionPolicy == EncryptionPolicy.RequireEncryption &&
+                    (!sslAuthenticationOptions.IsServer || (sslAuthenticationOptions.ApplicationProtocols != null && sslAuthenticationOptions.ApplicationProtocols.Count != 0));
+
+            if (sslAuthenticationOptions.CertificateContext != null && cacheSslContext)
+            {
+               sslAuthenticationOptions.CertificateContext.contexts.TryGetValue((int)sslAuthenticationOptions.EnabledSslProtocols, out sslCtx);
+            }
+
+            if (sslCtx == null)
+            {
+                // We did not get SslContext from cache
+                sslCtx = innerContext = AllocateSslContext(credential, sslAuthenticationOptions);
+            }
 
+            try
+            {
                 GCHandle alpnHandle = default;
                 try
                 {
+                    context = SafeSslHandle.Create(sslCtx, sslAuthenticationOptions.IsServer);
+                    Debug.Assert(context != null, "Expected non-null return value from SafeSslHandle.Create");
+                    if (context.IsInvalid)
+                    {
+                        context.Dispose();
```

Comment:
I assume this also ends up disposing of the sslCtx that was passed to it?

Reply:
no it should not. I updated the `finally` section to do so also if we don't cache.